{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Vg3hmGrGv6Le"
      },
      "source": [
        "\n",
        "![nufold.png](https://github.com/kiharalab/nufold/blob/master/nufold/NuFoldLogo.png?raw=1)</center>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Q5O5fhir3hvG"
      },
      "source": [
        "NuFold is a state-of-the-art method designed for predicting 3D RNA structures, leveraging deep learning for high accuracy and reliability. This tool is particularly useful for biologists and bioinformatics researchers focusing on RNA function and structure.\n",
        "\n",
        "License: GPL v3 for academic use. (For commercial use, please contact us for different licensing) Contact: Daisuke Kihara (dkihara@purdue.edu)\n",
        "\n",
        "Cite: [Kagaya, Y., Zhang, Z., Ibtehaz, N., Wang, X., Nakamura, T., Punuru, P.D., & Kihara, D. (2023). NuFold: A Novel Tertiary RNA Structure Prediction Method Using Deep Learning with Flexible Nucleobase Center Representation.](https://www.biorxiv.org/content/10.1101/2023.09.20.558715v1) In submission (an earlier version in bioRxiv).\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "koaiV0xzvMnr"
      },
      "source": [
        "# HOW TO USE NUFOLD\n",
        "- Below you will find a place to input your sequence and job name.\n",
        "- After giving sequence, go to the toolbar above, click `Runtime` and then `Run all`.\n",
        "- Change any settings you want to change before running."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AKtL8aghuged",
        "outputId": "f0afeab5-d96a-46a1-9c66-a0595228f50f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "RNA sequence length: 701\n",
            "RNA sequence saved as jobs/3UTR+IGR_base/3UTR+IGR/3UTR+IGR.fasta of length: 701\n"
          ]
        }
      ],
      "source": [
        "#@title Input your RNA sequence and provide a job name (optional)\n",
        "#@markdown # Add your data here\n",
        "import random\n",
        "import string\n",
        "import os\n",
        "\n",
        "def random_string(length=10):\n",
        "    \"\"\"Generate a random string of fixed length.\"\"\"\n",
        "    letters = string.ascii_lowercase\n",
        "    return ''.join(random.choice(letters) for i in range(length))\n",
        "\n",
        "rna_sequence = \"AGAGCAATTGTGTCAATTTTATTCAAACCTTAACACTCAGTCTTACAAATCATCACATTAAAACCCTAAGAAACGACTGCGGAATACAGAGTTGTACTTTCGCACCTTGAATTACATACGGTCAAAGCATATAACAACTTCTACGATCATCAAGCAGTTGTAAGTTAAATTATAAAAAaGCCTATAAATATATaAAGCTTTCTTTATCTTTATtGCtTGTGCtTGCTTAGTGTGtTAAATTTAAAaTAAGTGTGTTTAATTAAaGtTTGCTTTCTGTGTGTtGTGCTTAATAATAAACAAAATAACAAAAACAACAAAAAACATAAAATAAAAATAAAAGAAAAAATAAAATAAAATAAAATAAAATAAATAAGTAAAAATAAAAATAAAATAAAAACAAAAAACAAAAAACAAAAAACAAAACAAAACCCAAATTTGGCCAAAATTGTCCCTTTCGGGATCTTTTTGGGTTTTTGTTTTTTAATTTTTTTGTTGTTTTTATTTCATTTTTTGATTTTTGTTTGATTTTGTTTCATTTTTTGATTTTTGTTTCATTTTGATTTTATTTTTATTTTTATTTTTATTTTTATGGTTTCTATTAAACAGGAGAAATTCAAAAGAAATGACAAAACAGAGAAATAATTATGAGTAAAGAAAGAAAATAAACATAACATAATTAGGAAAAGCTGGACACGGCAAGA\"  # @param {type: \"string\"}\n",
        "length = len(rna_sequence)\n",
        "print(f\"RNA sequence length: {length}\")\n",
        "rna_sequence = rna_sequence.upper()\n",
        "job_name = '3UTR+IGR'  # @param {type: \"string\"}\n",
        "\n",
        "if not job_name:\n",
        "    job_name = random_string()\n",
        "\n",
        "# Prepare the FASTA content\n",
        "fasta_content = f\">{job_name}\\n{rna_sequence}\"\n",
        "jobs_dir = \"jobs\"\n",
        "if not os.path.exists(jobs_dir):\n",
        "    os.makedirs(jobs_dir)\n",
        "job_base_dir = os.path.join(jobs_dir, f\"{job_name}_base\")\n",
        "if not os.path.exists(job_base_dir):\n",
        "    os.makedirs(job_base_dir)\n",
        "actual_job_dir = os.path.join(job_base_dir, job_name)\n",
        "if not os.path.exists(actual_job_dir):\n",
        "    os.makedirs(actual_job_dir)\n",
        "fasta_filename = f\"{job_name}.fasta\"\n",
        "fasta_filepath = os.path.join(actual_job_dir, fasta_filename)\n",
        "with open(fasta_filepath, \"w\") as fasta_file:\n",
        "    fasta_file.write(fasta_content)\n",
        "\n",
        "print(f\"RNA sequence saved as {fasta_filepath} of length: {length}\")\n",
        "\n",
        "#@markdown If you want to download NuFold's output to your computer automatically, you can check this box here. Otherwise after the run is complete you can go to `/nufold_output` on the file explorer on the left side and download the folder with your job name.\n",
        "download = False #@param {type:\"boolean\"}\n",
        "\n",
        "#@markdown Additional Settings:\n",
        "num_recycles = 3 # @param {type:\"number\"}\n",
        "msa_speed = 2 # @param {type:\"slider\", min:0, max:2, step:1}\n",
        "#@markdown - 0 for slowest, 1 for normal, 2 for fast (Faster = Lower Quality)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5e7SHqCosh31",
        "outputId": "81861b01-1630-4eb7-b6d3-15d2f67bb6e4"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "rMSA installation complete.\n",
            "Libraries installation complete.\n",
            "ipknot installation complete.\n",
            "NuFold setup complete.\n",
            "All installations and setups are complete. Check logs.txt for details.\n"
          ]
        }
      ],
      "source": [
        "#@title Installing Dependencies (This is going to take a approximately 5 minutes when you run this notebook the first time)\n",
        "import os\n",
        "import subprocess\n",
        "\n",
        "!rm -rf sample_data\n",
        "\n",
        "def run_command(command, log_file, message=None):\n",
        "    with open(log_file, 'a') as file:\n",
        "        subprocess.run(command, shell=True, stdout=file, stderr=subprocess.STDOUT)\n",
        "    if message:\n",
        "        print(message)\n",
        "\n",
        "logs_file = 'logs.txt'\n",
        "open(logs_file, 'w').close()\n",
        "\n",
        "# Check and Install rMSA if not exists\n",
        "if not os.path.exists('rMSA'):\n",
        "    run_command(\n",
        "        \"apt-get install aria2 && \"\n",
        "        \"git clone https://github.com/pylelab/rMSA && \"\n",
        "        \"cd rMSA/database/ && \"\n",
        "        \"aria2c -q -R -x 16 -j 20 -s 65536 -c --optimize-concurrent-downloads https://kiharalab.org/nufold/database.zip && \"\n",
        "        \"unzip database.zip && \"\n",
        "        \"rm database.zip && \"\n",
        "        \"cd ../../\",\n",
        "        logs_file,\n",
        "        \"rMSA installation complete.\"\n",
        "    )\n",
        "else:\n",
        "    print(\"rMSA directory already exists. Skipping installation.\")\n",
        "\n",
        "\n",
        "\n",
        "# Install libraries\n",
        "run_command(\n",
        "    \"pip install ml-collections dm-tree deepspeed protobuf scipy biopython py3dmol\",\n",
        "    logs_file,\n",
        "    \"Libraries installation complete.\"\n",
        ")\n",
        "\n",
        "# Check and Install ipknot if not exists\n",
        "if not os.path.exists('/content/ipknot'):\n",
        "    run_command(\n",
        "        \"wget https://github.com/satoken/ipknot/releases/download/v1.1.0/ipknot-1.1.0-x86_64-linux.zip && \"\n",
        "        \"unzip ipknot-1.1.0-x86_64-linux.zip && \"\n",
        "        \"rm ipknot-1.1.0-x86_64-linux/README.md && \"\n",
        "        \"mv ipknot-1.1.0-x86_64-linux/ipknot /content && \"\n",
        "        \"rmdir ipknot-1.1.0-x86_64-linux && \"\n",
        "        \"rm ipknot-1.1.0-x86_64-linux.zip && \"\n",
        "        \"chmod +x /content/ipknot\",\n",
        "        logs_file,\n",
        "        \"ipknot installation complete.\"\n",
        "    )\n",
        "else:\n",
        "    print(\"ipknot already exists. Skipping installation.\")\n",
        "\n",
        "\n",
        "# Check and Setup NuFold if not exists\n",
        "# Make sure to add gitclone nufold here.\n",
        "if not os.path.exists('/content/run_nufold.py'):\n",
        "    run_command(\n",
        "        \"git clone https://github.com/kiharalab/NuFold &&\"\n",
        "        \"mv /content/NuFold/nufold /content && \"\n",
        "        \"mv /content/NuFold/run_nufold.py /content && \"\n",
        "        \"rm -rf /content/NuFold\",\n",
        "        logs_file,\n",
        "        \"NuFold setup complete.\"\n",
        "    )\n",
        "else:\n",
        "    print(\"NuFold is already set up. Skipping installation.\")\n",
        "\n",
        "print(\"All installations and setups are complete. Check logs.txt for details.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KDs4o5Bv35MI",
        "outputId": "8da3cf76-989e-4829-eec8-5548a0e5af15",
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/jobs/3UTR+IGR_base/3UTR+IGR\n",
            "ipknot output saved\n",
            "/content\n",
            "creating /tmp//rMSA_5623_067865265\n",
            ">/content/jobs/3UTR+IGR_base/3UTR+IGR/3UTR+IGR.fasta\tlength=701\n",
            "AGAGCAATTGTGTCAATTTTATTCAAACCTTAACACTCAGTCTTACAAATCATCACATTAAAACCCTAAGAAACGACTGCGGAATACAGAGTTGTACTTTCGCACCTTGAATTACATACGGTCAAAGCATATAACAACTTCTACGATCATCAAGCAGTTGTAAGTTAAATTATAAAAAAGCCTATAAATATATAAAGCTTTCTTTATCTTTATTGCTTGTGCTTGCTTAGTGTGTTAAATTTAAAATAAGTGTGTTTAATTAAAGTTTGCTTTCTGTGTGTTGTGCTTAATAATAAACAAAATAACAAAAACAACAAAAAACATAAAATAAAAATAAAAGAAAAAATAAAATAAAATAAAATAAAATAAATAAGTAAAAATAAAAATAAAATAAAAACAAAAAACAAAAAACAAAAAACAAAACAAAACCCAAATTTGGCCAAAATTGTCCCTTTCGGGATCTTTTTGGGTTTTTGTTTTTTAATTTTTTTGTTGTTTTTATTTCATTTTTTGATTTTTGTTTGATTTTGTTTCATTTTTTGATTTTTGTTTCATTTTGATTTTATTTTTATTTTTATTTTTATTTTTATGGTTTCTATTAAACAGGAGAAATTCAAAAGAAATGACAAAACAGAGAAATAATTATGAGTAAAGAAAGAAAATAAACATAACATAATTAGGAAAAGCTGGACACGGCAAGA\n",
            "using 2 cpu\n",
            "==== rfam pre-screening ====\n",
            "RF01577\n",
            "RF02409\n",
            "cp /content/rMSA/database/rfam_annotations.tsv.gz /tmp/rMSA_5623_067865265/db0to1\n",
            "rm /tmp/rMSA_5623_067865265/db0to1\n",
            "cut -f1 /tmp/rMSA_5623_067865265/rfam1.tab|sort|uniq > /tmp/rMSA_5623_067865265/rfam1.0.list\n",
            "split -l 5000 /tmp/rMSA_5623_067865265/rfam1.0.list /tmp/rMSA_5623_067865265/rfam1.0.list.split.\n",
            "/content/rMSA/bin/blastdbcmd -db /content/rMSA/database/rnacentral.fasta -entry_batch /tmp/rMSA_5623_067865265/rfam1.0.list.split.aa -out /tmp/rMSA_5623_067865265/rfam1.0.db.aa\n",
            "sort -k4g /tmp/rMSA_5623_067865265/rfam1.tab | head -200000 | /content/rMSA/bin/trimBlastN /tmp/rMSA_5623_067865265/rfam1.0.db.aa - 701 /tmp/rMSA_5623_067865265/rfam1.0.trim.split.aa\n",
            "rm /tmp/rMSA_5623_067865265/rfam1.0.db.aa /tmp/rMSA_5623_067865265/rfam1.0.list.split.aa\n",
            "cat /tmp/rMSA_5623_067865265/rfam1.0.trim.split.* | /content/rMSA/bin/fasta2pfam - | sort -u -k2 | /content/rMSA/bin/pfam2fasta - > /tmp/rMSA_5623_067865265/rfam1.0.db\n",
            "rm  /tmp/rMSA_5623_067865265/rfam1.0.trim.split.*\n",
            "cat /tmp/rMSA_5623_067865265/rfam1.*.db > /tmp/rMSA_5623_067865265/rfam1.db\n",
            "cp /content/rMSA/database/Rfam.full_region.gz /tmp/rMSA_5623_067865265/db0to2\n",
            "rm /tmp/rMSA_5623_067865265/db0to2\n",
            "cat /tmp/rMSA_5623_067865265/rfam2.*.db > /tmp/rMSA_5623_067865265/rfam2.db\n",
            "cat: '/tmp/rMSA_5623_067865265/rfam2.*.db': No such file or directory\n",
            "cat /tmp/rMSA_5623_067865265/rfam1.db /tmp/rMSA_5623_067865265/rfam2.db > /tmp/rMSA_5623_067865265/db0\n",
            "cat /tmp/rMSA_5623_067865265/db0 | gzip - > /tmp/rMSA_5623_067865265/db0.gz\n",
            "cp /tmp/rMSA_5623_067865265/db0.gz /content/jobs/3UTR+IGR_base/3UTR+IGR/3UTR+IGR.db0.gz\n",
            "  116 /tmp/rMSA_5623_067865265/rfam1.0.db\n",
            "  116 /tmp/rMSA_5623_067865265/rfam1.db\n",
            "    0 /tmp/rMSA_5623_067865265/rfam2.db\n",
            "  116 /tmp/rMSA_5623_067865265/db0\n",
            "  348 total\n",
            "==== blastn pre-screening ====\n",
            "/content/rMSA/bin/blastn -num_threads 2 -query /tmp/rMSA_5623_067865265/seq.fasta -strand plus -db /content/rMSA/database/rnacentral.fasta -out /tmp/rMSA_5623_067865265/blastn0.tab -task blastn -max_target_seqs 50000 -outfmt '6 saccver sstart send evalue bitscore nident staxids'\n",
            "cut -f1 /tmp/rMSA_5623_067865265/blastn0.tab|sort|uniq > /tmp/rMSA_5623_067865265/blastn0.list\n",
            "split -l 5000 /tmp/rMSA_5623_067865265/blastn0.list /tmp/rMSA_5623_067865265/blastn0.list.split.\n",
            "/content/rMSA/bin/blastdbcmd -db /content/rMSA/database/rnacentral.fasta -entry_batch /tmp/rMSA_5623_067865265/blastn0.list.split.aa -out /tmp/rMSA_5623_067865265/blastn0.db.aa\n",
            "sort -k4g /tmp/rMSA_5623_067865265/blastn0.tab | head -200000 | /content/rMSA/bin/trimBlastN /tmp/rMSA_5623_067865265/blastn0.db.aa - 701 /tmp/rMSA_5623_067865265/blastn0.trim.split.aa\n",
            "rm /tmp/rMSA_5623_067865265/blastn0.db.aa /tmp/rMSA_5623_067865265/blastn0.list.split.aa\n",
            "cat /tmp/rMSA_5623_067865265/blastn0.trim.split.* | /content/rMSA/bin/fasta2pfam - | sort -u -k2 | /content/rMSA/bin/pfam2fasta - > /tmp/rMSA_5623_067865265/blastn0.db\n",
            "rm  /tmp/rMSA_5623_067865265/blastn0.trim.split.*\n",
            "cat /tmp/rMSA_5623_067865265/db0 /tmp/rMSA_5623_067865265/blastn*.db > /tmp/rMSA_5623_067865265/trim.db\n",
            "/content/rMSA/bin/cd-hit-est-2d -T 2 -i /tmp/rMSA_5623_067865265/seq.fasta -i2 /tmp/rMSA_5623_067865265/trim.db -c 1 -o /tmp/rMSA_5623_067865265/cdhitest2d.db -l 280 -M 5000\n",
            "================================================================\n",
            "Program: CD-HIT, V4.8.1 (+OpenMP), Sep 23 2019, 14:35:48\n",
            "Command: /content/rMSA/bin/cd-hit-est-2d -T 2 -i\n",
            "         /tmp/rMSA_5623_067865265/seq.fasta -i2\n",
            "         /tmp/rMSA_5623_067865265/trim.db -c 1 -o\n",
            "         /tmp/rMSA_5623_067865265/cdhitest2d.db -l 280 -M 5000\n",
            "\n",
            "Started: Sat Mar  8 13:01:29 2025\n",
            "================================================================\n",
            "                            Output                              \n",
            "----------------------------------------------------------------\n",
            "total seq in db1: 1\n",
            "total seq in db2: 132\n",
            "longest and shortest : 701 and 701\n",
            "Total letters: 701\n",
            "Sequences have been sorted\n",
            "longest and shortest : 1542 and 290\n",
            "Total letters: 135627\n",
            "\n",
            "Approximated minimal memory consumption:\n",
            "Sequence        : 0M\n",
            "Buffer          : 2 X 12M = 24M\n",
            "Table           : 2 X 16M = 33M\n",
            "Miscellaneous   : 4M\n",
            "Total           : 62M\n",
            "\n",
            "Table limit with the given memory limit:\n",
            "Max number of representatives: 4000000\n",
            "Max number of word counting entries: 617169410\n",
            "\n",
            "..........        0  compared          0  clusters\n",
            "\n",
            "132 compared\t0 clustered\n",
            "writing non-redundant sequences from db2\n",
            "writing clustering information\n",
            "program completed !\n",
            "\n",
            "Total CPU time 0.16\n",
            "/content/rMSA/bin/cd-hit-est -T 2 -i /tmp/rMSA_5623_067865265/cdhitest2d.db -c 1 -o /tmp/rMSA_5623_067865265/db -l 280 -M 5000\n",
            "================================================================\n",
            "Program: CD-HIT, V4.8.1 (+OpenMP), Sep 23 2019, 14:35:48\n",
            "Command: /content/rMSA/bin/cd-hit-est -T 2 -i\n",
            "         /tmp/rMSA_5623_067865265/cdhitest2d.db -c 1 -o\n",
            "         /tmp/rMSA_5623_067865265/db -l 280 -M 5000\n",
            "\n",
            "Started: Sat Mar  8 13:01:29 2025\n",
            "================================================================\n",
            "                            Output                              \n",
            "----------------------------------------------------------------\n",
            "total seq: 132\n",
            "longest and shortest : 1542 and 290\n",
            "Total letters: 135627\n",
            "Sequences have been sorted\n",
            "\n",
            "Approximated minimal memory consumption:\n",
            "Sequence        : 0M\n",
            "Buffer          : 2 X 12M = 24M\n",
            "Table           : 2 X 16M = 33M\n",
            "Miscellaneous   : 4M\n",
            "Total           : 62M\n",
            "\n",
            "Table limit with the given memory limit:\n",
            "Max number of representatives: 3217473\n",
            "Max number of word counting entries: 617168582\n",
            "\n",
            "# comparing sequences from          0  to        132\n",
            "---------- new table with      122 representatives\n",
            "\n",
            "      132  finished        122  clusters\n",
            "\n",
            "Approximated maximum memory consumption: 63M\n",
            "writing new database\n",
            "writing clustering information\n",
            "program completed !\n",
            "\n",
            "Total CPU time 0.22\n",
            "rm /tmp/rMSA_5623_067865265/cdhitest2d.db\n",
            "cat /tmp/rMSA_5623_067865265/db | gzip - > /tmp/rMSA_5623_067865265/db.gz\n",
            "cp /tmp/rMSA_5623_067865265/db.gz /content/jobs/3UTR+IGR_base/3UTR+IGR/3UTR+IGR.db.gz\n",
            "rm /content/jobs/3UTR+IGR_base/3UTR+IGR/3UTR+IGR.db0.gz\n",
            "==== making covariance model from local alignment of 122 sequences by nhmmer ====\n",
            "/content/rMSA/bin/qnhmmer --noali -A /tmp/rMSA_5623_067865265/nhmmer.a2m --cpu 2 --watson /tmp/rMSA_5623_067865265/seq.fasta /tmp/rMSA_5623_067865265/db | grep 'no alignment saved'\n",
            "/content/rMSA/bin/a3m2msa /tmp/rMSA_5623_067865265/nhmmer.a2m | grep -ohP '^\\S+' | /content/rMSA/bin/fastaNA - > /tmp/rMSA_5623_067865265/nhmmer.unfilter.afa.tmp\n",
            "echo 701 != 352. realign by clustalo\n",
            "/content/rMSA/bin/clustalo --p1=/tmp/rMSA_5623_067865265/seq.fasta --p2=/tmp/rMSA_5623_067865265/nhmmer.unfilter.afa.tmp --is-profile --threads=2 | /content/rMSA/bin/RemoveNonQueryPosition - | /content/rMSA/bin/fixAlnX - N /tmp/rMSA_5623_067865265/nhmmer.unfilter.afa\n",
            "rm /tmp/rMSA_5623_067865265/nhmmer.unfilter.afa.tmp\n",
            "/content/rMSA/bin/hhfilter -i /tmp/rMSA_5623_067865265/nhmmer.unfilter.afa -id 99 -cov 50 -o /tmp/rMSA_5623_067865265/nhmmer.afa\n",
            "hhlib=/content/rMSA/bin/..\n",
            "Input file = /tmp/rMSA_5623_067865265/nhmmer.unfilter.afa\n",
            "Output file = /tmp/rMSA_5623_067865265/nhmmer.afa\n",
            "Read /tmp/rMSA_5623_067865265/nhmmer.unfilter.afa with 4 sequences\n",
            "Alignment in /tmp/rMSA_5623_067865265/nhmmer.unfilter.afa contains 701 match states\n",
            "1 out of 4 sequences passed filter (50% min coverage, 99% max pairwise sequence identity)\n",
            "Writing A3M alignment to /tmp/rMSA_5623_067865265/nhmmer.afa\n",
            "Done\n",
            "/content/rMSA/bin/fasta2pfam /tmp/rMSA_5623_067865265/nhmmer.afa |grep -ohP '\\S+$' | cat -n | grep -ohP '\\d+\\s+\\S+'|sed 's/^/>/g'|sed 's/\\t/\\n/g' > /tmp/rMSA_5623_067865265/nhmmer.fas\n",
            "/content/rMSA/bin/reformat.pl fas sto /tmp/rMSA_5623_067865265/nhmmer.fas /tmp/rMSA_5623_067865265/nhmmer.sto\n",
            "Use of uninitialized value $1 in printf at /content/rMSA/bin/reformat.pl line 750.\n",
            "Reformatted /tmp/rMSA_5623_067865265/nhmmer.fas with 1 sequence from fas to sto and written to file /tmp/rMSA_5623_067865265/nhmmer.sto\n",
            "/content/rMSA/bin/RNAfold --noPS /tmp/rMSA_5623_067865265/seq.fasta | awk '{print $1}' | tail -n +3 > /tmp/rMSA_5623_067865265/RNAfold.dbn\n",
            "cp /tmp/rMSA_5623_067865265/RNAfold.dbn /content/jobs/3UTR+IGR_base/3UTR+IGR/3UTR+IGR.dbn\n",
            "cp /tmp/rMSA_5623_067865265/RNAfold.dbn /tmp/rMSA_5623_067865265/RNAfold.gap.dbn\n",
            "/content/rMSA/bin/cmbuild --hand -F /tmp/rMSA_5623_067865265/infernal.cm /tmp/rMSA_5623_067865265/RNAfold.gap.sto\n",
            "# cmbuild :: covariance model construction from multiple sequence alignments\n",
            "# INFERNAL 1.1.3 (Nov 2019)\n",
            "# Copyright (C) 2019 Howard Hughes Medical Institute.\n",
            "# Freely distributed under the BSD open source license.\n",
            "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
            "# CM file:                                            /tmp/rMSA_5623_067865265/infernal.cm\n",
            "# alignment file:                                     /tmp/rMSA_5623_067865265/RNAfold.gap.sto\n",
            "# overwrite CM file if necessary:                     yes\n",
            "# use #=GC RF annotation to define consensus columns: yes\n",
            "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
            "#                                                                      rel entropy\n",
            "#                                                                      -----------\n",
            "# idx    name                     nseq eff_nseq   alen  clen  bps bifs    CM   HMM description\n",
            "# ------ -------------------- -------- -------- ------ ----- ---- ---- ----- ----- -----------\n",
            "       1 RNAfold.gap                 1     0.77    701   701  209    7 0.589 0.294 E=0.0\n",
            "#\n",
            "# CPU time: 7.48u 0.02s 00:00:07.50 Elapsed: 00:00:07.62\n",
            "/content/rMSA/bin/cmcalibrate --cpu 2 /tmp/rMSA_5623_067865265/infernal.cm\n",
            "# cmcalibrate :: fit exponential tails for CM E-values\n",
            "# INFERNAL 1.1.3 (Nov 2019)\n",
            "# Copyright (C) 2019 Howard Hughes Medical Institute.\n",
            "# Freely distributed under the BSD open source license.\n",
            "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
            "# CM file:                                     /tmp/rMSA_5623_067865265/infernal.cm\n",
            "# number of worker threads:                    2 [--cpu]\n",
            "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
            "#\n",
            "# Calibrating CM(s):\n",
            "#\n",
            "#                        predicted                                                   actual   \n",
            "#                       running time              percent complete                running time\n",
            "# model name            (hr:min:sec)  [........25........50........75..........]  (hr:min:sec)\n",
            "# --------------------  ------------  ------------------------------------------  ------------\n",
            "  RNAfold.gap               04:53:49  [=========="
          ]
        }
      ],
      "source": [
        "#@title Data Preprocess: Compute pseudoknots and perform MSA\n",
        "#@markdown This step is going to take a while, allow some time for this step (keep the page open)\n",
        "\n",
        "# Paths\n",
        "jobs_dir = \"jobs\"\n",
        "job_dir = os.path.join(jobs_dir, job_name)\n",
        "fasta_filepath = os.path.join(job_dir, fasta_filename)\n",
        "ss_output_path = os.path.join(job_dir, f\"{job_name}.ss\")\n",
        "\n",
        "# Change to the job directory\n",
        "%cd {actual_job_dir}\n",
        "# Run ipknot and capture output\n",
        "!../../../ipknot {fasta_filename} > {fasta_filename.replace('.fasta', '')}.ipknot.ss\n",
        "print(\"ipknot output saved\")\n",
        "\n",
        "%cd ../../..\n",
        "# Run rMSA.pl\n",
        "seqpath = os.path.join(actual_job_dir, fasta_filename)\n",
        "!sed -i 's@$dbdir/nt@@g' rMSA/rMSA.pl\n",
        "!rMSA/rMSA.pl {seqpath} -cpu=2 -fast={msa_speed}\n",
        "print(\"rMSA.pl run completed.\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "woN1sckydWRP"
      },
      "outputs": [],
      "source": [
        "#@title Now we're going to setup and actually do the NuFold part\n",
        "\n",
        "# #@markdown Hang in there we're almost done\n",
        "\n",
        "\n",
        "if not os.path.exists('checkpoints'):\n",
        "    os.makedirs('checkpoints')\n",
        "    !wget -O checkpoints/global_step145245.pt http://kiharalab.org/nufold/global_step145245.pt\n",
        "else:\n",
        "    print('\"checkpoints\" already fetched')\n",
        "\n",
        "\n",
        "seq = os.path.join(actual_job_dir, fasta_filename)\n",
        "!mkdir nufold_output\n",
        "!python /content/run_nufold.py \\\n",
        "  --ckpt_path /content/checkpoints/global_step145245.pt \\\n",
        "  --input_fasta {seq} \\\n",
        "  --input_dir {job_base_dir}\\\n",
        "  --output_dir /content/nufold_output \\\n",
        "  --config_preset initial_training \\\n",
        "  --recycle {num_recycles}\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "orMwqpBrbf38"
      },
      "outputs": [],
      "source": [
        "#@title Display Result\n",
        "import os\n",
        "import py3Dmol\n",
        "from IPython.display import display, Markdown, Image\n",
        "from google.colab import files as colabdl\n",
        "import shutil\n",
        "\n",
        "# Assuming output_path and fasta_filename are defined as shown previously\n",
        "output_path = os.path.join(\"/content/nufold_output\", fasta_filename.replace(\".fasta\", ''))\n",
        "files = os.listdir(output_path)\n",
        "\n",
        "# Filter for the PDB files with the specific prefix and the PLDDT plot\n",
        "pdb_files = [x for x in files if x.startswith(job_name) and x.endswith(\".pdb\")]\n",
        "plddt_plot = f\"{job_name}_plddt_plot.png\"\n",
        "\n",
        "# Function to visualize PDB file with py3Dmol\n",
        "def visualize_pdb_file(pdb_file_path):\n",
        "    with open(pdb_file_path) as ifile:\n",
        "        system = \"\".join([x for x in ifile])\n",
        "    view = py3Dmol.view(width=400, height=300)  # Adjust size as needed\n",
        "    view.addModelsAsFrames(system)\n",
        "    view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':50,'max':90}}})  # Apply styling\n",
        "    view.zoomTo()\n",
        "    return view\n",
        "\n",
        "# If there's at least one PDB file that matches the criteria, visualize it\n",
        "if pdb_files:\n",
        "    display(Markdown(f\"# Your Results\"))\n",
        "    pdb_files.sort()\n",
        "    #@markdown Provide the number of ranked structures you want to view\n",
        "    num_structures = 1 # @param {type:\"integer\"}\n",
        "    num_structures = min(num_structures, len(pdb_files))\n",
        "\n",
        "    for pdb_file in pdb_files[:num_structures]:\n",
        "        display(Markdown(f\"## File Name: {pdb_file}\"))\n",
        "        view = visualize_pdb_file(os.path.join(output_path, pdb_file))\n",
        "        view.show()\n",
        "        if plddt_plot in files:\n",
        "            display(Markdown(f\"## PLDDT Plot\"))\n",
        "            display(Image(os.path.join(output_path, plddt_plot)))\n",
        "        else:\n",
        "            display(Markdown(\"PLDDT plot not found.\"))\n",
        "else:\n",
        "    display(Markdown(\"No PDB files with the specified prefix found.\"))\n",
        "\n",
        "# Code for downloading the ZIP file if needed\n",
        "zipname = fasta_filename.replace(\".fasta\", '')\n",
        "if download:\n",
        "    zip_filename = f\"{zipname}.zip\"\n",
        "    shutil.make_archive(zip_filename.replace('.zip', ''), 'zip', output_path)\n",
        "    display(Markdown(\"## The download is probably going to take a while.\"))\n",
        "    colabdl.download(zip_filename)  # This line will work in Google Colab to trigger the download"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "gpuType": "T4",
      "machine_shape": "hm",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}